### DISCUSSION TITLE
return first match of yaml containing multiple documents

### DISCUSSION THREAD

**Discussion by joshcoburn** [mikefarah/yq#1326] [Q&A]
Hi I have a simple yaml file, which contains multiple documents: https://raw.githubusercontent.com/sassoftware/viya4-deployment/main/roles/vdm/templates/resources/openldap.yaml

I'm simply trying to query the container image from the yaml.. and I first attempted this:

`curl -s https://raw.githubusercontent.com/sassoftware/viya4-deployment/main/roles/vdm/templates/resources/openldap.yaml | yq -N '.spec.template.spec.containers[0].image'`

which returns:
osixia/openldap:1.3.0
null

I realize it is returning null, because it is failing to find my query in the second document within this file.  Is there an easy way with yq to only return the string I'm after, without null?

**Comment by joshcoburn** [comment]
I forced it to work doing this: 

`curl -s https://raw.githubusercontent.com/sassoftware/viya4-deployment/main/roles/vdm/templates/resources/openldap.yaml | yq -N '.spec.template.spec.containers[0].image' | sed s/"null"// | xargs`

which returns:
osixia/openldap:1.3.0

But was hoping for a more elegant way...

**Comment by mikefarah** [comment]
Just pipe the result through a select:


```bash
curl -s https://raw.githubusercontent.com/sassoftware/viya4-deployment/main/roles/vdm/templates/resources/openldap.yaml | yq -N '.spec.template.spec.containers[0].image  | select(. != null)'
```

**Reply by joshcoburn** [reply]
Perfect, thank you!

**Comment by s6mike** [comment]
I have two non-null results from two docs, but I only want the first (the second is an overlapping data set but I want to prioritise results from the first when they are available).

Is there a simple way to only return the first result?

**Comment by s6mike** [comment]
I solved this by converting the results to JSON then capturing in a bash array and just returning the first result.

But then I came up with a different solution which I think is more elegant:
`yq '.query | select( . != null)' file1.yaml file2.yaml | yq 'select(document_index == 0)'

The first yq command will return a separate document for each file which matches, and then the second yq command will return only the result from the first document. If only one file matches, then it will always be the first document, so it will always be used for the result.

**Reply by rocketraman** [reply]
This works, but is there any way to do it with one `yq` command rather than piping to a separate `yq`? I was expecting `yq '.query | select( whatever ) | select(document_inex == 0)' file1.yaml file2.yaml` to work but it does not.

**Reply by mikefarah** [reply]
You can do with using 'eval-all' command. By default, 'eval' is used, which will iterate over the given files, and documents within the files, running the expression against each document independently (it's more memory efficient, so it doesn't load up everything in mem).

'eval-all/ea' on the other hand, loads everything into memory, and then runs the expression once.

```
yq ea '[.query | select( whatever )] | .[0]' file1.yaml file2.yaml
```

- Read all the documents, filter them and put the results into a single array
- Get the first matching element in the array
- If there are no matching elements, this will return null.

**Reply by rocketraman** [reply]
Thank you,that works!
